### DISCUSSION TITLE
Image-List with async loading from stream - Best practice?

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#14516] [Q&A] [answered]
I need guidance on the best way of solving my problem. I want to show a probably huge list of images, but I have to load the images from an S3-kind of storage. I can retrieve them via a Stream. But now the problems arise:

Fetching the available image-keys is an async call to my backend - this is quite trivial. But fetching the image-stream itself takes some time (obviously). And I was not successfull binding a stream to an Image-Element without converting it to a byte[]. And then there is no buffering if the same image should be shown again as the user scrolls through the list.

So the question is: what is the best and most performant way of rendering a huge collection of images that can only be fetched via a stream? Is there a best-practice or sample for this use case available?

**Comment by limeniye** [comment]
The first thing that comes to my mind is to use an ItemsRepeater with a unique identifier cache.
Check this ```IndexForUniqueId``` sample [there](https://learn.microsoft.com/en-us/windows/apps/design/controls/items-repeater#configure-the-data-source).

**Reply by TopperDEL** [reply]
I have a memory cache in place for the downloaded bytes. Not sure if caching the Images/UIElements themselves would be a better option. Would I gain much if I use ItemsRepeater instead of a GridView? My assumption is that loading the image-bytes asynch (see below) is the most time consuming task.

**Comment by jeromelaban** [comment]
It is possible to provide a stream to `BitmapImage`, but the stream is read synchronously (even the `SetSourceAsync` one) to its end at this time.

Still, one way to handle this is to have your own local cache, that is able to provide BitmapSource` instances that can be reused across listview items. Those can be provided empty at first, and filled when the image has been downloaded.

**Reply by TopperDEL** [reply, marked as answer]
This is the way I do it at the moment. I load my keys (which are strings) using an IncrementalLoadingCollection and provide that as ItemsSource to a GridView. The ItemTemplate is just an Image binding the source to that key. A converter grabs the Key, creates a BitmapImage and starts a Task to Load the image-bytes. It immediately returns the BitmapImage while the image gets loaded. If the image-bytes are available, I convert it to a MemorySteam and SetSource() it on the image:

```
public object? Convert(object value, Type targetType, object parameter, string language)
{
    if (_objectService == null)
    {
        return null;
    }
    if (value == null)
    {
        return null;
    }
    var image = new BitmapImage();
    _ = Task.Run(() => LoadImageAsync(value as string, image));
    return image;
}

private async Task LoadImageAsync(string key, BitmapImage image)
{
    try
    {
        var bytes = _memoryCache.Get<byte[]>(key);
        if (bytes == null)
        {
            var objectInfo = await _objectService.GetObjectAsync(_bucket, key);
            if (objectInfo.SystemMetadata.ContentLength > 0)
            {
                using (var downloadOperation = await _objectService.DownloadObjectAsync(_bucket, key, new DownloadOptions(), false))
                {
                    await downloadOperation.StartDownloadAsync();
                    if (downloadOperation.Completed)
                    {
                        _memoryCache.Set(key, downloadOperation.DownloadedBytes, DateTime.Now.AddMinutes(10));
                        bytes = downloadOperation.DownloadedBytes;
                    }
                }
            }
        }

        if (bytes != null)
        {
            _dispatcher.TryEnqueue(() =>
            {
                image.SetSource(new MemoryStream(bytes).AsRandomAccessStream());
            });
        }
    }
    catch
    {
    }
}
```
This works very good on Windows. On Android the UI still stutters while setting the Source of the BitmapImage even if I switch over to SetSourceAsync(). But I guess this is what you meant, Jerome, right?

**Reply by TopperDEL** [reply]
I have another small problem with the solution above: on Android the Image does not get invalidated after SetSource. It shows the image correctly when I scroll it out of view and come back. I would assume this is a bug?

**Reply by jeromelaban** [reply]
Android is not particularly good at dealing with synchronous operations. After thinking about it a bit more, you may be able to use the `file://` uris, where a off-dispatcher process can take charge of loading the files.

As for the images not reloading, yes, it's likely a bug, though we've never encountered it.

**Reply by TopperDEL** [reply]
Can you eloberate more what you mean with the file-Uri? Is there an Event or something where I could inject my Code?

The not-reloading does not happen on the Release build, for whatever reason. So I'm fine with it currently.

**Reply by jeromelaban** [reply]
Instead of using SetSource, use a local file that `Image` will use to load the content.
